lib/Dialect/Datapath/DatapathOps.cpp

```diff
@@ -59,6 +66,323 @@ static void printCompressFormat(OpAsmPrinter &printer, Operation *op,
           << resultTypes.size() << "]";
 }
 
+//===----------------------------------------------------------------------===//
+// Compressor Tree Logic.
+//===----------------------------------------------------------------------===//
+
+// Construct a full adder for three 1-bit inputs.
+std::pair<CompressorBit, CompressorBit>
+CompressorTree::fullAdderWithDelay(OpBuilder &builder, CompressorBit a,
+                                   CompressorBit b, CompressorBit c) {
+
+  auto aXorB = builder.createOrFold<comb::XorOp>(loc, a.val, b.val, true);
+  Value sum_val = builder.createOrFold<comb::XorOp>(loc, aXorB, c.val, true);
+
+  auto carry_val = builder.createOrFold<comb::OrOp>(
+      loc,
+      ArrayRef<Value>{
+          builder.createOrFold<comb::AndOp>(loc, a.val, b.val, true),
+          builder.createOrFold<comb::AndOp>(loc, aXorB, c.val, true)},
+      true);
+
+  auto sum_delay = std::max(std::max(a.delay, b.delay) + 1, c.delay) + 1;
+  auto carry_delay = sum_delay + 1;
+
+  CompressorBit sum = {sum_val, sum_delay};
+  CompressorBit carry = {carry_val, carry_delay};
+  std::pair<CompressorBit, CompressorBit> fa{sum, carry};
+  ++numFullAdders;
+  return fa;
+}
+
+// Construct a full adder for three 1-bit inputs.
+std::pair<CompressorBit, CompressorBit>
+CompressorTree::halfAdderWithDelay(OpBuilder &builder, CompressorBit a,
+                                   CompressorBit b) {
+  auto sum_val = builder.createOrFold<comb::XorOp>(loc, a.val, b.val, true);
+  auto carry_val = builder.createOrFold<comb::AndOp>(loc, a.val, b.val, true);
+
+  auto sum_delay = std::max(a.delay, b.delay) + 1;
+  auto carry_delay = sum_delay;
+
+  CompressorBit sum = {sum_val, sum_delay};
+  CompressorBit carry = {carry_val, carry_delay};
+  std::pair<CompressorBit, CompressorBit> ha{sum, carry};
+  return ha;
+}
+
+// Map input rows to column representation
+CompressorTree::CompressorTree(const SmallVector<SmallVector<Value>> &addends,
+                               Location loc)
+    : originalAddends(addends), usingTiming(false), numStages(0),
+      numFullAdders(0), loc(loc) {
+  assert(addends.size() > 2);
+  // Number of bits in a row == bitwidth of input addends
+  // Compressors will be formed of uniform bitwidth addends
+  width = addends[0].size();
+  SmallVector<SmallVector<CompressorBit>> initColumns(width);
+  columns = initColumns;
+  // Convert addends rows to columns
+  // Known bits analysis constructs a minimal array - skipping zeros
+  for (auto row : addends) {
```

Comment:
Can we initialize width and columns with constructors and add an assertion for width?
```suggestion
// Map input rows to column representation
CompressorTree::CompressorTree(size_t width, const SmallVector<SmallVector<Value>>& addends,
                               Location loc)
    : originalAddends(addends), usingTiming(false), numStages(0),
      numFullAdders(0), loc(loc), width(width), columns(width) {
  assert(addends.size() > 2);
  // Convert addends rows to columns
  // Known bits analysis constructs a minimal array - skipping zeros
  for (const auto& row : addends) {
    assert(row.size() == width);
```